### PR TITLE
[SPARK-28419][SQL] A patch for SparkThriftServer to support multi-tenant authentication for spark-2.4.0

### PR DESCRIPTION
**Pull request by AngersZhuuuu**
## What changes were proposed in this pull request?

This PR is about a patch to improve spark-2.4.0 to support  multi-tenant authentication.

In this PR, hope to discuss with Spark SQL developers about how to enable SparkThriftServer proxy user. And accept suggestions about this method and improve code structure.(ps: there are some unsightly or crude code.)

## How was this patch tested?

I have test it in our enviorment, it can read and write with proxy user in Kerberos.
Hope some one can test it in different enviorment.
Need more  concurrency  test.

### CODE REVIEW

**Comment by AngersZhuuuu** [comment]
gentle ping @wangyum @juliuszsompolski @dongjoon-hyun @turboFei @yaooqinn @maropu

File: sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkSQLSessionManager.scala [hunk 1]
```diff
@@ -52,22 +54,55 @@ private[hive] class SparkSQLSessionManager(hiveServer: HiveServer2, sqlContext:
...
-          delegationToken)
-    val session = super.getSession(sessionHandle)
+    var session: HiveSession = null
+    var sparkSession: SparkSession = null
+    if (withImpersonation) {
+      val sessionWithUGI = new HiveSessionImplwithUGI(protocol, username, passwd, hiveConf, ipAddress, delegationToken)
+      val ugi = sessionWithUGI.getSessionUgi
+      val originalCreds = ugi.getCredentials
+      val creds = new Credentials()
+      ThriftServerHadoopUtil.doAs(ugi)(() => hadoopTokenProvider.obtainDelegationTokens(creds, username))
```
Comment (AngersZhuuuu) [inline sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkSQLSessionManager.scala]: Should consider more about token expire.

**Comment by wangyum** [comment]
Could you make a PR against `master` branch? We usually merge into `master` first, then backport to `branch-2.4` and `branch-2.3` if needed.

**Comment by AngersZhuuuu** [comment]
> Could you make a PR against `master` branch? We usually merge into `master` first, then backport to `branch-2.4` and `branch-2.3` if needed.

I will do this. This is preparatory work .

**Comment by dongjoon-hyun** [comment]
Thank you for making a PR, @AngersZhuuuu .
However, we review a PR on `master` branch to guarantee not to make any regression.
So, I'll close this PR for now. You can reuse this PR by reopening after this is done in `master`.

**Comment by AngersZhuuuu** [comment]
> Thank you for making a PR, @AngersZhuuuu .
> However, we review a PR on `master` branch to guarantee not to make any regression.
> So, I'll close this PR for now. You can reuse this PR by reopening after this is done in `master`.

I will test more and make it more strong  to make a pr on master

**Comment by dongjoon-hyun** [comment]
Thank you for your understanding, @AngersZhuuuu !

**Pull request closed without merging**